accepted/future-releases/generalized-typedef-2018/feature-specification.md

```diff
@@ -0,0 +1,137 @@
+specification, with a proposed form in
+[this CL](https://dart-review.googlesource.com/c/sdk/+/81414).
+The following sections summarize the changes.
+
+
+### Syntax
+
+The grammar is modified as follows in order to support this feature:
+
+```
+typeAlias ::=
+  metadata 'typedef' typeAliasBody |
+  metadata 'typedef' identifier typeParameters? '=' type ';' // CHANGED
+```
+
+*The modification is that a type alias declaration of the form that uses
+`=` can define the type alias to denote any type, not just a function
+type.*
+
+
+### Static Analysis 
+
+There is no change in the treatment of existing syntax, so we describe only
+the treatment of syntactic forms that can be new when this feature is added.
+
+The effect of a non-generic type alias declaration of the form that uses
+`=` with name `F` is to bind the name `F` in the library scope of the
+enclosing library to the type denoted by the right hand side of `=` in that
+declaration.
+
+*This is not new, but it applies to a larger set of situations now that the
+right hand side can be any type. Let us call that type on the right hand
+side `T`. This means that `F` can be used as a type annotation, and the
+entity which is given type `F` is considered to have type `T`, with the
+same meaning as in the declaration of `F` (even if some declaration in the
+current library gives a new meaning to an identifier in `T`).*
+
+The effect of a generic type alias declaration of the form
+
+```dart
+typedef F<X1 extends B1 .. Xk extends Bk> = T;
+```
+
+where metadata is omitted but may be present, is to bind the name `F` in
+the library scope of the enclosing library to a mapping from type argument
+lists to types, such that a parameterized type of the form `F<T1..Tk>` is
+an application of that mapping that denotes the type `[T1/X1 .. Tk/Xk]T`.
+
+Let _F_ be a generic type alias of the form that uses `=` with type
+parameter declarations
+_X<sub>1</sub> extends B<sub>1</sub> .. X<sub>k</sub> extends B<sub>k</sub>_.
+It is a compile-time error unless satisfaction of the declared bounds
+_B<sub>1</sub> .. B<sub>k</sub>_ implies that the right hand side is
+regular-bounded, and all types that occur as subterms of the right hand
+side are well-bounded.
+
+Let `F<T1..Tn>` be a parameterized type where `F` denotes a declaration of
+the form
+```dart
+typedef F<X1 extends B1 .. Xk extends Bk> = T;
+```
+where metadata is omitted but may be present. It is a compile-time error if
+`n != k` and it is a compile-time error if `F<T1..Tn>` is not well-bounded.
+
+*These errors are not new, but they apply to a larger set of situations now
+that the right hand side can be any type.*
+
+When a `typeName` (*that is, `identifier` or `identifier.identifier`*)
+that resolves to a generic type alias declaration is used as a type or
+evaluated as an expression, it is subject to instantiation to bound.
+
+*This treatment of generic type aliases is again the same as it was
+previously, but it involves a larger set of types.*
+
```

Comment:
Maybe mention that type aliases still cannot be recursive. The type alias itself must not occur in the RHS of the definition.

Reply:
Indeed! Done.